src/annotator/index.js

```diff
@@ -68,28 +68,32 @@ function init() {
     sidebar = new Sidebar(document.body, eventBus, guest, getConfig('sidebar'));
 
     // Expose sidebar window reference for use by same-origin guest frames.
-    window_.__hypothesis.sidebarWindow = sidebar.iframe.contentWindow;
+    window_.__hypothesis.sidebarWindow = sidebar.ready.then(
+      () => sidebar.iframe.contentWindow
+    );
   }
 
   // Clear `annotations` value from the notebook's config to prevent direct-linked
   // annotations from filtering the threads.
   const notebook = new Notebook(document.body, eventBus, getConfig('notebook'));
 
   // Set up communication between this host/guest frame and the sidebar frame.
-  const sidebarWindow = sidebar
-    ? sidebar.iframe.contentWindow
-    : /** @type {HypothesisWindow} */ (window.parent).__hypothesis
-        ?.sidebarWindow;
+  const sidebarWindow =
+    window_.__hypothesis.sidebarWindow ||
+    /** @type {HypothesisWindow} */ (window.parent).__hypothesis?.sidebarWindow;
   if (sidebarWindow) {
-    guest.crossframe.connectToSidebar(sidebarWindow);
+    const sidebarOrigin = new URL(sidebarLinkElement.href).origin;
+    sidebarWindow.then(frame =>
+      guest.crossframe.connectToSidebar(frame, sidebarOrigin)
+    );
   } else {
     // eslint-disable-next-line no-console
     console.warn(
```

Comment:
This condition shouldn't be reached in normal usage, but could be reached if the Hypothesis client was manually loaded into a cross-origin guest frame with `{ subFrameIdentifier: '...' }` in the configuration. It could also be reached if somehow Hypothesis was unloaded in the host frame before it had loaded in the guest frame.

Reply:
I think on the first scenario, a cross-origin guest, this condition won't be reached because it will throw an error earlier trying to to access (window.parent).__hypothesis.

I didn't thought about the second scenario, Hypothesis unloaded in the host frame before it had loaded in the guests frame.

Reply:
> I think on the first scenario, a cross-origin guest, this condition won't be reached because it will throw an error earlier trying to to access (window.parent).__hypothesis.

You're right. I thought that `window.parent` would be a different kind of `Window` object in a cross-origin context that would just appear to have fewer properties and not expose any globals, so `window.parent.__hypothesis` would return `undefined`.

However this is not what happens. It does indeed generate an exception:

<img width="1029" alt="window-parent-access" src="https://user-images.githubusercontent.com/2458/127309776-a9d206d9-c7cc-43a0-866b-7dcd8b2d04bd.png">

Reply:
I pushed a commit to improve this. See https://github.com/hypothesis/client/pull/3611#issuecomment-888216575.